### PR TITLE
feat(button): make children required

### PR DESCRIPTION
**Pull request by alexanderpolyankin**
[FX-1501]

### Description

Make `children` of `Button` required.

### How to test

- `yarn typecheck`

### Screenshots

| Before.                                 | After.                                  |
| --------------------------------------- | --------------------------------------- |
| Insert screenshots or screen recordings | Insert screenshots or screen recordings |

### Review

- [ ] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/api-principles.md)
- [ ] Annotate all `props` in component with documentation
- [ ] Create `examples` for component
- [ ] Ensure that deployed demo has expected results and good examples
- [ ] Ensure that tests pass by running `yarn test`
- [ ] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [ ] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

[FX-1501]: https://toptal-core.atlassian.net/browse/FX-1501

### CODE REVIEW

File: packages/shared/src/styles/add-classes-to-children.tsx [hunk 1]
```diff
@@ -1,8 +1,9 @@
-import React, { ComponentType, ReactNode } from 'react'
+import React, { ReactNode } from 'react'
 
 import { Classes } from './types'
 
-type ConfigItem = [ComponentType, Classes]
+type ConfigType = unknown
```
Comment (alexanderpolyankin) [inline packages/shared/src/styles/add-classes-to-children.tsx:5]: The problem is that a component with required `children` is not `ComponentType` or `FC` anymore, because in these types `children` are optional.

Comment (teimurjan) [reply packages/shared/src/styles/add-classes-to-children.tsx:5]: I'd avoid using `unknown`. Can you describe this problem in detail, please? Do you have a TS error after making `children` required?

File: packages/shared/src/styles/add-classes-to-children.tsx [hunk 2]
```diff
@@ -19,8 +20,8 @@ export default ({ classes, children, config }: Arguments) => {
 
     let childResult = childNode
 
-    config(classes).forEach(([componentType, configClasses]) => {
-      if (childNode.type === componentType) {
+    config(classes).forEach(([configType, configClasses]) => {
```
Comment (teimurjan) [inline packages/shared/src/styles/add-classes-to-children.tsx:23]: Why did you need this renaming?

Comment (alexanderpolyankin) [reply packages/shared/src/styles/add-classes-to-children.tsx:23]: Because the provided type is no longer a `ComponentType`, I'd like to indicate that the function works on any any provided types and just components.

**Comment by alexanderpolyankin** [comment]
> I'd avoid using unknown. Can you describe this problem in detail, please? Do you have a TS error after making children required?

I added a comment about this in the PR. The reason that a component with required `children` is not assignable to `ComponentType` or `FC` because these types add `{ children?: ReactNode }`, which is not required.

**Comment by havenchyk** [comment]
What about the point from ticket

> Hint. Consider extracting ButtonBase component and use it in Circular and rectangular buttons.

File: packages/picasso/src/ButtonGroup/ButtonGroup.tsx [hunk 3]
```diff
@@ -25,33 +19,33 @@ const useStyles = makeStyles<Theme>(styles, {
...
-    }
-  ]
-]
+const useChildrenWithClasses = (classes: Classes, children?: ReactNode) => {
+  return useMemo(
+    () =>
+      addClassesToChildren({
+        children,
+        classes,
+        config: () => [
```
Comment (alexanderpolyankin) [inline packages/picasso/src/ButtonGroup/ButtonGroup.tsx:28]: The change is required to avoid using `typeof Button` instead of `ComponentType`.

**Comment by alexanderpolyankin** [comment]
> What about the point from ticket
> 
> > Hint. Consider extracting ButtonBase component and use it in Circular and rectangular buttons.

It was not required to make the `children` required. Do you want me to extract `ButtonBase` anyway?

File: packages/shared/src/styles/add-classes-to-children.tsx [hunk 4]
```diff
@@ -1,17 +1,21 @@
...
-export interface Arguments {
+export interface Arguments<T> {
   classes: Classes
   children: ReactNode
-  config: Config
+  config: Config<T>
 }
 
-export default ({ classes, children, config }: Arguments) => {
+const addClassesToChildren = <T extends unknown>({
```
Comment (havenchyk) [inline packages/shared/src/styles/add-classes-to-children.tsx:14]: is `extends unknown` because of jsx?

File: packages/picasso/src/ButtonGroup/ButtonGroup.tsx (hunk 3 above)
Comment (havenchyk) [reply packages/picasso/src/ButtonGroup/ButtonGroup.tsx:28]: could you please explain how is it different?

File: packages/picasso-forms/src/SubmitButton/SubmitButton.tsx [hunk 5]
```diff
@@ -16,11 +17,12 @@ type CircularButtonTypeProps = {
 type ActionButtonTypeProps = {
   buttonType: 'action'
   variant?: null
+  children?: ReactNode
 }
 
 // Intersection with the type { id?: string } is needed here because of
 // TS compiler issue https://github.com/microsoft/TypeScript/issues/34793
-export type Props = Omit<ButtonProps, 'type' | 'variant'> & {
+export type Props = Omit<ButtonProps, 'type' | 'variant' | 'children'> & {
```
Comment (havenchyk) [inline packages/picasso-forms/src/SubmitButton/SubmitButton.tsx:25]: reason?

Comment (alexanderpolyankin) [reply packages/picasso-forms/src/SubmitButton/SubmitButton.tsx:25]: Because `children` are required in `ButtonProps`, but not all type of `SubmitButton` have required children.

Comment (havenchyk) [reply packages/picasso-forms/src/SubmitButton/SubmitButton.tsx:25]: 👍

**Comment by alexanderpolyankin** [comment]
@havenchyk @teimurjan Found a better approach with `ElementType`, so generics are not required. Please have a look.

**Review by havenchyk — APPROVED** [review]

**Review by vedrani — APPROVED** [review]

**Review by teimurjan — APPROVED** [review]
Great job! 👍
